### ISSUE TITLE
Jest has detected the following 1 open handle potentially keeping Jest from exiting: CustomGC

### ISSUE THREAD

**Issue by ctsstc** [prisma/prisma#18146]
### Bug description

When I run `jest --detectOpenHandles` I get the following:
(Which seems to stem from the Prisma Client)

```
Jest has detected the following 1 open handle potentially keeping Jest from exiting:

  ●  CustomGC

      at Runtime._loadModule (../node_modules/jest-runtime/build/index.js:1111:29)
      at load (../node_modules/@prisma/client/runtime/library.js:73:326)
      at ../node_modules/@prisma/client/runtime/library.js:73:759
      at runInChildSpan (../node_modules/@prisma/client/runtime/library.js:70:25817)
      at Fr.loadLibrary (../node_modules/@prisma/client/runtime/library.js:73:677)
      at Vt.loadEngine (../node_modules/@prisma/client/runtime/library.js:101:557)
      at Vt.instantiateLibrary (../node_modules/@prisma/client/runtime/library.js:100:1478)
```

### How to reproduce

```
npx jest --detectOpenHandles
```

### Expected behavior

There should be no open handles left over from Jest tests to prisma. There should be no open DB handles anyways since the tests are not connecting a database.

### Prisma information

Cannot share on the given project unfortunately.

### Environment & setup

- OS: macOS
- Database: PostgreSQL
- Node.js version: v16.15.1

### Prisma Version

```
prisma                  : 4.11.0
@prisma/client          : 4.11.0
Current platform        : darwin-arm64
Query Engine (Node-API) : libquery-engine 8fde8fef4033376662cad983758335009d522acb (at node_modules/@prisma/engines/libquery_engine-darwin-arm64.dylib.node)
Migration Engine        : migration-engine-cli 8fde8fef4033376662cad983758335009d522acb (at node_modules/@prisma/engines/migration-engine-darwin-arm64)
Format Wasm             : @prisma/prisma-fmt-wasm 4.11.0-57.8fde8fef4033376662cad983758335009d522acb
Default Engines Hash    : 8fde8fef4033376662cad983758335009d522acb
Studio                  : 0.483.0
```

**Comment by thomasmarren** [comment]
Also seeing this. 

It looks like it was introduced in `v4.9.0`. Downgrading to `v4.8.1` removes the message.

**Comment by cbix** [comment]
Worth noting this happens with
```
generator client {
  provider = "prisma-client-js"
}
```
and `prismaClient.$disconnect()` being called in the `afterAll` callback. According to the [docs](https://www.prisma.io/docs/concepts/components/prisma-client/working-with-prismaclient/connection-management#disconnect),
> When you call `$disconnect()`, Prisma Client [...] ends the Query Engine child process and closes all connections

From the jest error it seems that unloading the query engine library is not working properly.

**Comment by aravind605** [comment]
> Also seeing this.
> 
> It looks like it was introduced in `v4.9.0`. Downgrading to `v4.8.1` removes the message.
Jest still exiting causing test suite to fail even after downgrading to 4.8.1

**Comment by thomasmarren** [comment]
@aravind605 I'm not sure your issue is directly related to this. 

At least in my case, the open handle message does not cause the test suite to fail, it's just a warning when using `--detectOpenHandles`.

**Comment by aravind605** [comment]
> @aravind605 I'm not sure your issue is directly related to this.
> 
> At least in my case, the open handle message does not cause the test suite to fail, it's just a warning when using `--detectOpenHandles`.

In my case this is my jest config 

module.exports = {
  preset: "ts-jest",
  testEnvironment: "node",
  displayName: {
    name: "node-express",
    color: "greenBright",
  },
  testTimeout: 120000,
  verbose: true,
  detectOpenHandles: true,
  collectCoverage: true,
  forceExit: true,
  moduleDirectories: ["node_modules", "src"]
};

I'm using detect open handles flag and observe that every test case succeed but at the end suite fails which I believe prisma client is not gracefully closing down and still attempting to search for active connection after the container stopped. 

![image](https://user-images.githubusercontent.com/26151395/223772207-c1519ef5-b1ef-4001-b4df-6fc8ec1b4823.png)

**Comment by james-julius** [comment]
Also seeing this - downgrading to `4.8.1` makes no difference to me

**Comment by andrewmclagan** [comment]
Can confirm that `^4.11.0` causes open handle issue. Downgrading to `4.8.1` removes the issue.

**Comment by andrewmclagan** [comment]
I created a repository to reproduce the issue: https://github.com/andrewmclagan/prisma-open-handles

It seems it only happens when using pnpm and could possibly be related to [this issue](https://github.com/prisma/prisma/issues/13672), although I think likely its another issue. As this only starts happening on version `^4.11.0`

**Comment by ctsstc** [comment]
@andrewmclagan  I'm using `npm` and see the issue.

**Comment by mehdibo** [comment]
Same problem here. I tried downgrading to `4.8.1` and calling `$disconnect` after all tests but still the same issue.

**Comment by ImLunaHey** [comment]
Using `npm` myself and hitting this so it's definitely not just pnpm.

**Comment by aravind605** [comment]
I've observed this might concern only when we attempt to write integration tests with prisma/client library. But not when the lib is mocked.

**Comment by rodrigocipriani** [comment]
I'm having the same issue here. And it's causing Garbage Collector not to work well and so Heap Memory is huge after some test files. I've tested 4.8.1 and 4.11.0.

**Comment by luizamoorim** [comment]
I have the same issue and the Heap size is increasing a lot after writing some tests files like @rodrigocipriani has mentioned. 
I tested lots of approaches to close this warning and decrease memory consumption with Jest + Prisma but without success.

**Comment by alexey2baranov** [comment]
it's not just minor issue. it brokes our pipeline on current project. probably by timeout.

**Comment by GadigaeARUN** [comment]
> it's not just minor issue. it brokes our pipeline on current project. probably by timeout.

true. It breaks our pipeline on test stage.

**Comment by aravind605** [comment]
> it's not just minor issue. it brokes our pipeline on current project. probably by timeout.

true it breaks pipeline. Anyone got it successfully passing the bdd tests? please let me know the version you have?

**Comment by onadeemdev** [comment]
same issue, had to downgrade to 4.7.0 for it to work

**Comment by rodrigocipriani** [comment]
> 

Here 4.7.0 didn't work. Same problem.

**Comment by avery-sungage** [comment]
We also see this bug on 4.3.1

**Comment by rossmotley** [comment]
I suspect this might be an issue with one of the dev dependencies of prisma. I encountered the issue after doing a an `npm update` which updated `prisma` and `@prisma/client` from `4.9.0` to `4.12.0`. 

However, after updating just the prisma packages with `npm install @prisma/client@4.12.0 prisma@4.12.0`, I am not seeing the issue.

**Comment by ctsstc** [comment]
I think tests will always exit/finish, I think there's a timeout/threshold for how long a test can take before it skips and continues. This is likely why you see the error, but it will always finish.

I do not have `bullmq` as a dependency but this is a Nest JS project.

**Comment by rodrigocipriani** [comment]
Libraries I'm using:

```json
"dependencies": {
    "@opentelemetry/api": "^1.3.0",
    "@opentelemetry/auto-instrumentations-node": "^0.36.0",
    "@opentelemetry/exporter-trace-otlp-http": "^0.35.0",
    "@opentelemetry/sdk-node": "^0.35.0",
    "@prisma/client": "^4.11.0",
    "@splitsoftware/splitio": "^10.21.1",
    "@types/express-rate-limit": "^5.1.3",
    "@types/rate-limit-redis": "^1.7.4",
    "@types/redis": "^2.8.32",
    "assert-plus": "^1.0.0",
    "cls-hooked": "^4.2.2",
    "dotenv": "^16.0.3",
    "express": "^4.17.3",
    "express-rate-limit": "^5.3.0",
    "kafkajs": "^2.0.2",
    "lusca": "^1.7.0",
    "module-alias": "^2.2.2",
    "mongodb": "^4.13.0",
    "node-fetch": "^2.6.7",
    "pg": "^8.7.3",
    "rate-limit-redis": "^2.1.0",
    "redis": "^3.1.2",
    "tsoa": "^5.1.1",
    "winston": "^3.6.0"
  },
  "devDependencies": {
    "@faker-js/faker": "^7.1.0",
    "@guidecx/config-prettier": "^1.0.0",
    "@types/assert-plus": "^1.0.4",
    "@types/cls-hooked": "^4.3.3",
    "@types/cors": "^2.8.12",
    "@types/eslint": "^7.28.0",
    "@types/express": "^4.17.6",
    "@types/jest": "^29.4.2",
    "@types/lusca": "^1.7.1",
    "@types/node": "^16.18.16",
    "@types/node-fetch": "^2.6.1",
    "@types/supertest": "^2.0.11",
    "@typescript-eslint/eslint-plugin": "^5.48.0",
    "@typescript-eslint/parser": "^5.48.0",
    "concurrently": "^7.0.0",
    "copyfiles": "^2.4.1",
    "cross-env": "^7.0.3",
    "docker-compose": "^0.23.17",
    "dotenv-cli": "^6.0.0",
    "eslint": "^7.32.0",
    "eslint-config-prettier": "^8.3.0",
    "husky": "^7.0.4",
    "jest": "^29.5.0",
    "lint-staged": "^12.3.7",
    "nodemon": "^2.0.12",
    "prettier": "2.3.2",
    "prisma": "^4.11.0",
    "supertest": "^6.1.4",
    "ts-jest": "^29.0.5",
    "ts-node": "^10.9.1",
    "typescript": "^4.9.4",
    "uuid-random": "^1.3.2"
  }
```

**Comment by dengelke** [comment]
Managed to get ours to work with 4.8.1

**Comment by rodrigocipriani** [comment]
Didn't work here also with 4.8.1.

I've created a very simple example you can replicate.

`package.json`
```json
{
  "name": "prismatest",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "test": "jest --detectOpenHandles"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "dependencies": {
    "@prisma/client": "^4.12.0",
    "prisma": "^4.12.0"
  },
  "devDependencies": {
    "@types/jest": "^29.5.0",
    "jest": "^29.5.0",
    "ts-jest": "^29.1.0",
    "typescript": "^5.0.3"
  }
}
```

`jest.config.js``

```js
/** @type {import('ts-jest').JestConfigWithTsJest} */
module.exports = {
  preset: "ts-jest",
  testEnvironment: "node",
};
```

`tsconfig.json``

```json
{
  "compilerOptions": {
    "esModuleInterop": true
  }
}
```

`schema.prisma`

```
datasource db {
    url      = env("DATABASE_URL")
    provider = "postgresql"
}

generator client {
    provider = "prisma-client-js"
}

model User {
    id        Int      @id @default(autoincrement())
    createdAt DateTime @default(now())
    email     String   @unique
    name      String?
}
```

`sum.ts`

```typescript
function sum(a: number, b: number): number {
  return a + b;
}

export default sum;

```

`sum.test.ts`

```typescript
import { PrismaClient } from "@prisma/client";
import sum from "./sum";

const prismaClient = new PrismaClient();

test("adds 1 + 2 to equal 3", () => {
  expect(sum(1, 2)).toBe(3);
});
```

`npm test` 

```terminal
➜ prismatest npm test

> prismatest@1.0.0 test
> jest --detectOpenHandles

 PASS  ./sum.test.ts
  ✓ adds 1 + 2 to equal 3 (4 ms)

Test Suites: 1 passed, 1 total
Tests:       1 passed, 1 total
Snapshots:   0 total
Time:        1.332 s
Ran all test suites.

Jest has detected the following 1 open handle potentially keeping Jest from exiting:

  ●  CustomGC

      at Runtime._loadModule (node_modules/jest-runtime/build/index.js:1009:29)
      at load (node_modules/@prisma/client/runtime/library.js:73:326)
      at node_modules/@prisma/client/runtime/library.js:73:759
      at runInChildSpan (node_modules/@prisma/client/runtime/library.js:70:25817)
      at qr.loadLibrary (node_modules/@prisma/client/runtime/library.js:73:677)
      at zt.loadEngine (node_modules/@prisma/client/runtime/library.js:101:557)
      at zt.instantiateLibrary (node_modules/@prisma/client/runtime/library.js:100:1477)
```

**Comment by ctsstc** [comment]
Thanks for digging in to create a minimum reproduction @rodrigocipriani 

With rolling prisma back so far, I have to wonder if this has always been going on for the most part, but maybe Jest updated something that does a better job of catching this problem, so now we've been seeing it.

**Comment by rodrigocipriani** [comment]
> Thanks for digging in to create a minimum reproduction @rodrigocipriani 
> 
> With rolling prisma back so far, I have to wonder if this has always been going on for the most part, but maybe Jest updated something that does a better job of catching this problem, so now we've been seeing it.

Maybe you're right, now it's making better job catching problems. But GC still not working, and memory heap size just increases each test. But only when Prisma is "in the game". 
Right now I'm not sure that I've tested without Typescript. I I'll get some time tomorrow to test without Typescript.

**Comment by gitnlsn** [comment]
I faced the same issue with customGC, preventing jest from exiting. Downgrading to 4.8.1 removed the issue.

My dependencies and devDependencies are the following.

```json
  "dependencies": {
    "@prisma/client": "4.8.1",
    "express": "^4.18.2",
    "zod": "^3.21.4",
    "zod-express-middleware": "^1.4.0"
  },
  "devDependencies": {
    "@types/express": "^4.17.17",
    "@types/jest": "^29.5.0",
    "@types/supertest": "^2.0.12",
    "jest": "^29.5.0",
    "prisma": "4.8.1",
    "supertest": "^6.3.3",
    "ts-jest": "^29.1.0",
    "typescript": "^5.0.3"
  },
```

**Comment by rodrigocipriani** [comment]
> ```json
> 4.8.1
> ```

Hey, can you send the other files? I've just tested right now with 4.8.1 and still failing.

![image](https://user-images.githubusercontent.com/1211294/230617063-fb12f9fc-50fb-43d4-94d7-26a09f88cd5a.png)

**Comment by rodrigocipriani** [comment]
> Thanks for digging in to create a minimum reproduction @rodrigocipriani
> 
> With rolling prisma back so far, I have to wonder if this has always been going on for the most part, but maybe Jest updated something that does a better job of catching this problem, so now we've been seeing it.

I've just tested without Typescript and still not having CustomGC error. But the Heap size decreased a lot.

![image](https://user-images.githubusercontent.com/1211294/230617524-0bec9582-33f1-4a07-845c-f0ad3e350758.png)

**Comment by rodrigocipriani** [comment]
I've just published it on GitHub for who is interested

https://github.com/rodrigocipriani/prismatest

**Comment by gitnlsn** [comment]
> I've just published it on GitHub for who is interested
> 
> https://github.com/rodrigocipriani/prismatest

I've checked it. Ive cloned it and executed it. I found no issues when running sum.test.*.

But the issue appears when i establish a connection and make a create operation with prisma.

When I downgrade to version 4.8.1. The issue indeed disapppears.

https://github.com/gitnlsn/prisma-test

Also, I guess the issue might not be related to `prisma` version at devDependencies, but related to `@prisma/client` at dependencies.

I hope those information help to debug it.

**Comment by petrokravets** [comment]
4.8.1 didn't help..
Does anyone have any solution?

**Comment by ctsstc** [comment]
@janpio What do we need to get the "unconfirmed" label removed? Is there any additional information required? Others have added example repositories as an example for reproduction. Does confirmation need to happen from an internal team?

**Comment by janpio** [comment]
Yes. If you can confirm using one of the repositories that were posted, let us know - that makes it easier for us.

**Comment by aravind605** [comment]
4.8.1 didn't help me either!

**Comment by South-Paw** [comment]
Downgrading to `4.8.1` worked for me, make sure that you don't keep a ^ in front when downgrading...

```json
{
  "dependencies": {
    "@prisma/client": "4.8.1",
    "prisma": "^4.8.1"
  }
}
```

However, I have gotten `4.13.0` working now 🥳

I noted that my integration tests that use the jest `--runInBand` flag and have a `setupFilesAfterEnv` script that disconnects after each test didn't have the warnings - I assume it something to do with jest running tests in parallel and prisma not being able to clean up connections properly

```ts
// setupFilesAfterEnv script with --runInBand
afterEach(async () => {
  /** After each test, close database connections */
  await prisma.$disconnect();
});
```

I tried adding this afterEach to my unit tests without the `--runInBand` flag and still encountered the same problem though.

Ended up resolving it by ensuring prisma is mocked correctly in my unit tests (which... it should of already been but because of some changes I made it wasn't. But unit tests shouldn't actually use a prisma client - thats not a unit test 😆)

**Comment by rodrigocipriani** [comment]
The tests I made with:

- NodeJS 20
- Prisma 4.13.0
- on schema.prisma using the new `previewFeatures = ["jsonProtocol"]`

seems to solve the Heap Memory problem.

I'm not sure if NodeJS <= 18 will solve it, I haven't tested enough
Everything is working with the previewFeature, but I'm afraid of using it in production.

**Comment by aravind605** [comment]
Prisma 4.13.0 and Prisma 4.13.0 stable versions are released. 
Still no luck with jest tests exiting with a failure as mentioned in this comment. :( 

> > Thanks for digging in to create a minimum reproduction @rodrigocipriani
> > With rolling prisma back so far, I have to wonder if this has always been going on for the most part, but maybe Jest updated something that does a better job of catching this problem, so now we've been seeing it.
> 
> I've just tested without Typescript and still not having CustomGC error. But the Heap size decreased a lot.
> 
> ![image](https://user-images.githubusercontent.com/1211294/230617524-0bec9582-33f1-4a07-845c-f0ad3e350758.png)

**Comment by rodrigocipriani** [comment]
> Prisma 4.13.0 and Prisma 4.13.0 stable versions are released. 
> Still no luck with jest tests exiting with a failure as mentioned in this comment. :( 
> 
> > > Thanks for digging in to create a minimum reproduction @rodrigocipriani
> > > With rolling prisma back so far, I have to wonder if this has always been going on for the most part, but maybe Jest updated something that does a better job of catching this problem, so now we've been seeing it.
> > 
> > I've just tested without Typescript and still not having CustomGC error. But the Heap size decreased a lot.
> > 
> > ![image](https://user-images.githubusercontent.com/1211294/230617524-0bec9582-33f1-4a07-845c-f0ad3e350758.png)
> 
> 

Hey @aravind605 ... I didn't try to update this repo to test it. I did it in my other repo and it worked. 
Did you try to use the preview feature I cote in this comment?

https://github.com/prisma/prisma/issues/18146#issuecomment-1519227942

If I have some time I'll update the test repo you've mentioned.

Let me know if you tested with the json way.

**Comment by aravind605** [comment]
> > Prisma 4.13.0 and Prisma 4.13.0 stable versions are released.
> > Still no luck with jest tests exiting with a failure as mentioned in this comment. :(
> > > > Thanks for digging in to create a minimum reproduction @rodrigocipriani
> > > > With rolling prisma back so far, I have to wonder if this has always been going on for the most part, but maybe Jest updated something that does a better job of catching this problem, so now we've been seeing it.
> > > 
> > > 
> > > I've just tested without Typescript and still not having CustomGC error. But the Heap size decreased a lot.
> > > ![image](https://user-images.githubusercontent.com/1211294/230617524-0bec9582-33f1-4a07-845c-f0ad3e350758.png)
> 
> Hey @aravind605 ... I didn't try to update this repo to test it. I did it in my other repo and it worked. Did you try to use the preview feature I cote in this comment?
> 
> [#18146 (comment)](https://github.com/prisma/prisma/issues/18146#issuecomment-1519227942)
> 
> If I have some time I'll update the test repo you've mentioned.
> 
> Let me know if you tested with the jsonProtocol.

It might be due to the previous node modules. once I've cleared and re-installed
Issue is not appearing any more!! Thanks @rodrigocipriani and team prisma!

**Comment by rodrigocipriani** [comment]
@aravind605 I'm glad that now it's working for you, thanks for reaching us back.
I will not say that this "bug" is fixed because it's just working with a preview feature, so it's not so safe to use for now.

**Comment by kane1997** [comment]
Jest has detected the following 1 open handle potentially keeping Jest from exiting:

  ●  CustomGC

      at Runtime._loadModule (node_modules/jest-runtime/build/index.js:1111:29)
      at DefaultLibraryLoader.loadLibrary (node_modules/@prisma/client/runtime/index.js:24348:29)
      at LibraryEngine.loadEngine (node_modules/@prisma/client/runtime/index.js:24658:24)
      at LibraryEngine.instantiateLibrary (node_modules/@prisma/client/runtime/index.js:24613:5)
      at LibraryEngine.getConfig (node_modules/@prisma/client/runtime/index.js:24822:5)
      at PrismaClient._getActiveProvider (node_modules/@prisma/client/runtime/index.js:32650:30)



same for prisma version 4.9.0

**Comment by ctsstc** [comment]
Confirmed on `4.14.1` this does not look to be happening any more 🎉

**Comment by MarcusHSmith** [comment]
I'm reproducing in `4.15.0`
